File: backend/app/models/reserva_recurrente.py
"""
RESERVIVES - Modelo de Reserva Recurrente.

Almacena el patrón de recurrencia de una reserva de espacio.
Requiere aprobación administrativa antes de generar instancias.
Cada instancia aprobada se materializa como una ReservaEspacio normal.
"""

import enum
import uuid
from datetime import date, datetime

from sqlalchemy import Date, DateTime, Enum, ForeignKey, Integer, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class TipoRecurrencia(str, enum.Enum):
    SEMANAL = "SEMANAL"
    QUINCENAL = "QUINCENAL"
    MENSUAL = "MENSUAL"


class EstadoReservaRecurrente(str, enum.Enum):
    PENDIENTE_APROBACION = "PENDIENTE_APROBACION"
    APROBADA = "APROBADA"
    RECHAZADA = "RECHAZADA"
    CANCELADA = "CANCELADA"


class ReservaRecurrente(Base):
    """
    Patrón de reserva periódica que el admin debe aprobar.
    Una vez aprobada, el scheduler genera instancias (ReservaEspacio)
    automáticamente según el tipo de recurrencia hasta fecha_fin_recurrencia.
    """
    __tablename__ = "reservas_recurrentes"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    usuario_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("usuarios.id", ondelete="CASCADE"), nullable=False
    )
    espacio_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("espacios.id", ondelete="CASCADE"), nullable=False
    )
    tramo_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tramos_horarios.id", ondelete="CASCADE"), nullable=False
    )
    tipo_recurrencia: Mapped[TipoRecurrencia] = mapped_column(
        Enum(TipoRecurrencia, name="tipo_recurrencia", create_type=False),
        nullable=False,
    )
    fecha_inicio: Mapped[date] = mapped_column(Date, nullable=False)
    fecha_fin_recurrencia: Mapped[date] = mapped_column(Date, nullable=False)
    estado: Mapped[EstadoReservaRecurrente] = mapped_column(
        Enum(EstadoReservaRecurrente, name="estado_reserva_recurrente", create_type=False),
        nullable=False,
        default=EstadoReservaRecurrente.PENDIENTE_APROBACION,
    )
    observaciones: Mapped[str | None] = mapped_column(Text)
    motivo_rechazo: Mapped[str | None] = mapped_column(Text)
    tokens_por_instancia: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    # Fecha de la última instancia generada para este patrón
    ultima_instancia_generada: Mapped[date | None] = mapped_column(Date, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    usuario = relationship("Usuario")
    espacio = relationship("Espacio")
    tramo = relationship("TramoHorario")
    instancias = relationship(
        "ReservaEspacio",
        back_populates="reserva_recurrente",
        foreign_keys="ReservaEspacio.reserva_recurrente_id",
    )

    def __repr__(self) -> str:
        return f"<ReservaRecurrente {self.id} - {self.tipo_recurrencia.value} - {self.estado.value}>"

File: frontend/lib/models/reserva_recurrente.dart
/// Modelo de Reserva Recurrente.
///
/// Representa un patrón de reserva periódica que requiere aprobación
/// del administrador antes de generar instancias concretas.
library;

enum TipoRecurrencia {
  semanal('SEMANAL'),
  quincenal('QUINCENAL'),
  mensual('MENSUAL');

  final String value;
  const TipoRecurrencia(this.value);

  factory TipoRecurrencia.fromString(String v) =>
      TipoRecurrencia.values.firstWhere((e) => e.value == v,
          orElse: () => TipoRecurrencia.semanal);

  String get label {
    switch (this) {
      case TipoRecurrencia.semanal:
        return 'Semanal';
      case TipoRecurrencia.quincenal:
        return 'Quincenal';
      case TipoRecurrencia.mensual:
        return 'Mensual';
    }
  }
}

enum EstadoReservaRecurrente {
  pendienteAprobacion('PENDIENTE_APROBACION'),
  aprobada('APROBADA'),
  rechazada('RECHAZADA'),
  cancelada('CANCELADA');

  final String value;
  const EstadoReservaRecurrente(this.value);

  factory EstadoReservaRecurrente.fromString(String v) =>
      EstadoReservaRecurrente.values.firstWhere((e) => e.value == v,
          orElse: () => EstadoReservaRecurrente.pendienteAprobacion);
}

class ReservaRecurrente {
  final String id;
  final String usuarioId;
  final String espacioId;
  final String tramoId;
  final TipoRecurrencia tipoRecurrencia;
  final DateTime fechaInicio;
  final DateTime fechaFinRecurrencia;
  final EstadoReservaRecurrente estado;
  final String? observaciones;
  final String? motivoRechazo;
  final int tokensPorInstancia;
  final DateTime? ultimaInstanciaGenerada;
  final String? nombreUsuario;
  final String? nombreEspacio;
  final String? nombreTramo;
  final DateTime createdAt;
  final DateTime updatedAt;

  const ReservaRecurrente({
    required this.id,
    required this.usuarioId,
    required this.espacioId,
    required this.tramoId,
    required this.tipoRecurrencia,
    required this.fechaInicio,
    required this.fechaFinRecurrencia,
    required this.estado,
    this.observaciones,
    this.motivoRechazo,
    required this.tokensPorInstancia,
    this.ultimaInstanciaGenerada,
    this.nombreUsuario,
    this.nombreEspacio,
    this.nombreTramo,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ReservaRecurrente.fromJson(Map<String, dynamic> json) {
    return ReservaRecurrente(
      id: json['id'] as String,
      usuarioId: json['usuario_id'] as String,
      espacioId: json['espacio_id'] as String,
      tramoId: json['tramo_id'] as String,
      tipoRecurrencia: TipoRecurrencia.fromString(json['tipo_recurrencia'] as String),
      fechaInicio: DateTime.parse(json['fecha_inicio'] as String),
      fechaFinRecurrencia: DateTime.parse(json['fecha_fin_recurrencia'] as String),
      estado: EstadoReservaRecurrente.fromString(json['estado'] as String),
      observaciones: json['observaciones'] as String?,
      motivoRechazo: json['motivo_rechazo'] as String?,
      tokensPorInstancia: json['tokens_por_instancia'] as int,
      ultimaInstanciaGenerada: json['ultima_instancia_generada'] != null
          ? DateTime.parse(json['ultima_instancia_generada'] as String)
          : null,
      nombreUsuario: json['nombre_usuario'] as String?,
      nombreEspacio: json['nombre_espacio'] as String?,
      nombreTramo: json['nombre_tramo'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  bool get isPendiente => estado == EstadoReservaRecurrente.pendienteAprobacion;
  bool get isAprobada => estado == EstadoReservaRecurrente.aprobada;
  bool get isRechazada => estado == EstadoReservaRecurrente.rechazada;
  bool get isCancelada => estado == EstadoReservaRecurrente.cancelada;
  bool get isActiva => isAprobada || isPendiente;
}
